File: app/src/main/java/com/printf/kidsteacher/adapter/VideoAdapter.java
package com.printf.kidsteacher.adapter;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.TextView;

import com.printf.kidsteacher.R;
import com.printf.kidsteacher.activity.BaseActivity;
import com.printf.kidsteacher.activity.VideoActivity;
import com.printf.kidsteacher.activity.VideoDetailActivity;
import com.printf.kidsteacher.been.videoData.Datum;
import com.printf.kidsteacher.other.Ease;
import com.printf.kidsteacher.other.EasingInterpolator;
import com.squareup.picasso.Picasso;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class VideoAdapter extends RecyclerView.Adapter
{
    Context context;
    List<Datum> list;
    ArrayList<Datum> arraylist;
    Animation animation;
    boolean isOpened = false;

    public VideoAdapter(Context activity, List<Datum> data)
    {
        context = activity;
        list = data;
        arraylist = new ArrayList<>();
        arraylist.addAll(list);
    }

    public void filter(String charText)
    {
        list.clear();////main list
        if (charText.length() == 0) {
            list.addAll(arraylist);
        }
        else
        {
            for (Datum wp : arraylist)
            {
                if (wp.getVideoName().toLowerCase().contains(charText.toLowerCase()))
                {
                    list.add(wp);
                }
            }
        }
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i)
    {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_video,viewGroup,false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int i)
    {
        if(viewHolder instanceof MyViewHolder)
        {
            MyViewHolder holder = (MyViewHolder)viewHolder;
            holder.view.setVisibility(View.VISIBLE);
            holder.iv_temp.setVisibility(View.VISIBLE);
            holder.iv_img.setScaleType(ImageView.ScaleType.CENTER_CROP);
            Picasso.with(context).load(list.get(i).getThumbnailUrl()).into(holder.iv_img);
            if(list.get(i).getVideoName() != null && !list.get(i).getVideoName().equalsIgnoreCase(""))
            {
                holder.tv_name.setText(list.get(i).getVideoName());
                holder.tv_name.setTextSize(context.getResources().getDimension(R.dimen._6sdp));
                holder.tv_name.setGravity(Gravity.START);
                holder.tv_name.setMaxLines(2);
                holder.tv_name.setEllipsize(TextUtils.TruncateAt.END);
            }
        }
    }

    @Override
    public int getItemCount() {
        return list.size();
    }


    private class MyViewHolder extends RecyclerView.ViewHolder
    {
        ImageView iv_img,iv_temp;
        TextView tv_name;
        View view;

        public MyViewHolder(@NonNull final View itemView)
        {
            super(itemView);

            tv_name = itemView.findViewById(R.id.tv_name);
            iv_img = itemView.findViewById(R.id.iv_img);
            view = itemView.findViewById(R.id.view);
            iv_temp = itemView.findViewById(R.id.iv_temp);

            animation = AnimationUtils.loadAnimation(context,R.anim.bounce);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (isOpened)
                        return;
                    ObjectAnimator animator = ObjectAnimator.ofFloat(itemView, "translationY", 0, 50, 0);
                    animator.setInterpolator(new EasingInterpolator(Ease.BOUNCE_IN_OUT));
                    animator.setStartDelay(0);
                    animator.setDuration(1500);
                    animator.start();
                    isOpened = true;
                    animator.addListener(new Animator.AnimatorListener() {
                        @Override
                        public void onAnimationStart(Animator animator) {

                        }

                        @Override
                        public void onAnimationEnd(Animator animator) {
                            isOpened = false;
                            Intent intent = new Intent(context, VideoDetailActivity.class);
                            Bundle bundle = new Bundle();
                            bundle.putSerializable("list",(Serializable) list);
                            bundle.putInt("position",getAdapterPosition());
                            intent.putExtra("list",bundle);
                            context.startActivity(intent);
                            ((Activity)context).overridePendingTransition(R.anim.enter_left,R.anim.exit_right);
                        }

                        @Override
                        public void onAnimationCancel(Animator animator) {

                        }

                        @Override
                        public void onAnimationRepeat(Animator animator) {

                        }
                    });
                }
            });
        }
    }
}

File: app/src/main/java/com/printf/kidsteacher/adapter/VideoAdapter.kt
package com.printf.kidsteacher.adapter

import android.animation.Animator
import android.animation.ObjectAnimator
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.printf.kidsteacher.R
import com.printf.kidsteacher.activity.VideoDetailActivity
import com.printf.kidsteacher.been.videoData.Datum
import com.printf.kidsteacher.databinding.ItemVideoBinding
import com.printf.kidsteacher.other.Ease
import com.printf.kidsteacher.other.EasingInterpolator
import java.io.Serializable
import java.util.*
import kotlin.collections.ArrayList

class VideoAdapter(var context: Context, var list: ArrayList<Datum>) :
        RecyclerView.Adapter<VideoAdapter.MyViewHolder>() {
    var arraylist: ArrayList<Datum> = ArrayList()
    var isOpened = false
    fun filter(charText: String) {
        list.clear()
        if (charText.isEmpty()) {
            list.addAll(arraylist)
        } else {
            for (wp in arraylist) {
                if (wp.videoName!!.toLowerCase().contains(charText.toLowerCase())) {
                    list.add(wp)
                }
            }
        }
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): VideoAdapter.MyViewHolder {
        val layoutInflater = LayoutInflater.from(viewGroup.context)
        val binding = DataBindingUtil.inflate(
                layoutInflater,
                R.layout.item_video,
                viewGroup,
                false
        ) as ItemVideoBinding
        return MyViewHolder(binding)
    }

    override fun onBindViewHolder(viewHolder: VideoAdapter.MyViewHolder, i: Int) {
        val dataModel = list[i]
        viewHolder.bind(dataModel, i)
    }

    override fun getItemCount(): Int {
        return list.size
    }

    inner class MyViewHolder(var itemRowBinding: ItemVideoBinding) : RecyclerView.ViewHolder(itemRowBinding.root) {

        fun bind(data: Datum, position: Int) {
            itemRowBinding.item = data
            itemRowBinding.executePendingBindings()

            itemRowBinding.root.setOnClickListener {

                if (isOpened) return@setOnClickListener

                isOpened = false
                val animator = ObjectAnimator.ofFloat(itemView, "translationY", 0f, 50f, 0f)
                animator.interpolator = EasingInterpolator(Ease.BOUNCE_IN_OUT)
                animator.startDelay = 0
                animator.duration = 1500
                animator.start()

                animator.addListener(object : Animator.AnimatorListener {
                    override fun onAnimationStart(animator: Animator) {}
                    override fun onAnimationEnd(animator: Animator) {
                        isOpened = false
                        val intent = Intent(context, VideoDetailActivity::class.java)
                        val bundle = Bundle()
                        bundle.putSerializable("list", list as Serializable)
                        bundle.putInt("position", position)
                        intent.putExtra("list", bundle)
                        context.startActivity(intent)
                        (context as Activity).overridePendingTransition(R.anim.enter_left, R.anim.exit_right)
                    }

                    override fun onAnimationCancel(animator: Animator) {
                        isOpened = false
                    }

                    override fun onAnimationRepeat(animator: Animator) {}
                })
            }
        }

    }

    init {
        arraylist.addAll(list)
    }
}